File: python/dataproc_templates/mongo/mongo_to_gcs.py
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Sequence, Optional, Any
from logging import Logger
import argparse
import pprint

from pyspark.sql import SparkSession, DataFrameWriter

from dataproc_templates import BaseTemplate
from dataproc_templates.util.argument_parsing import add_spark_options
from dataproc_templates.util.dataframe_writer_wrappers import persist_dataframe_to_cloud_storage
import dataproc_templates.util.template_constants as constants


__all__ = ['MongoToGCSTemplate']


class MongoToGCSTemplate(BaseTemplate):
    """
    Dataproc template implementing exports from Mongo to Cloud Storage
    """

    @staticmethod
    def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
        parser: argparse.ArgumentParser = argparse.ArgumentParser()

        parser.add_argument(
            f'--{constants.MONGO_GCS_INPUT_URI}',
            dest=constants.MONGO_GCS_INPUT_URI,
            required=True,
            help='MONGO Cloud Storage Input Connection Uri'
        )
        parser.add_argument(
            f'--{constants.MONGO_GCS_INPUT_DATABASE}',
            dest=constants.MONGO_GCS_INPUT_DATABASE,
            required=True,
            help='MONGO Cloud Storage Input Database Name'
        )
        parser.add_argument(
            f'--{constants.MONGO_GCS_INPUT_COLLECTION}',
            dest=constants.MONGO_GCS_INPUT_COLLECTION,
            required=True,
            help='MONGO Cloud Storage Input Collection Name'
        )
        parser.add_argument(
            f'--{constants.MONGO_GCS_OUTPUT_FORMAT}',
            dest=constants.MONGO_GCS_OUTPUT_FORMAT,
            required=True,
            help='Output file format (one of: avro,parquet,csv,json)',
            choices=[
                constants.FORMAT_AVRO,
                constants.FORMAT_PRQT,
                constants.FORMAT_CSV,
                constants.FORMAT_JSON
            ]
        )
        parser.add_argument(
            f'--{constants.MONGO_GCS_OUTPUT_LOCATION}',
            dest=constants.MONGO_GCS_OUTPUT_LOCATION,
            required=True,
            help='Cloud Storage location for output files'
        )
        parser.add_argument(
            f'--{constants.MONGO_GCS_OUTPUT_MODE}',
            dest=constants.MONGO_GCS_OUTPUT_MODE,
            required=False,
            default=constants.OUTPUT_MODE_APPEND,
            help=(
                'Output write mode '
                '(one of: append,overwrite,ignore,errorifexists) '
                '(Defaults to append)'
            ),
            choices=[
                constants.OUTPUT_MODE_OVERWRITE,
                constants.OUTPUT_MODE_APPEND,
                constants.OUTPUT_MODE_IGNORE,
                constants.OUTPUT_MODE_ERRORIFEXISTS
            ]
        )
        add_spark_options(parser, constants.get_csv_output_spark_options("mongo.gcs.output."))

        known_args: argparse.Namespace
        known_args, _ = parser.parse_known_args(args)

        return vars(known_args)

    def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:

        logger: Logger = self.get_logger(spark=spark)

        # Arguments
        input_uri: str = args[constants.MONGO_GCS_INPUT_URI]
        input_database: str = args[constants.MONGO_GCS_INPUT_DATABASE]
        input_collection: str = args[constants.MONGO_GCS_INPUT_COLLECTION]
        output_format: str = args[constants.MONGO_GCS_OUTPUT_FORMAT]
        output_mode: str = args[constants.MONGO_GCS_OUTPUT_MODE]
        output_location: str = args[constants.MONGO_GCS_OUTPUT_LOCATION]

        ignore_keys = {constants.MONGO_GCS_INPUT_URI}
        filtered_args = {key:val for key,val in args.items() if key not in ignore_keys}
        logger.info(
            "Starting Mongo to Cloud Storage Spark job with parameters:\n"
            f"{pprint.pformat(filtered_args)}"
        )

        # Read
        input_data = spark.read \
            .format(constants.FORMAT_MONGO) \
            .option(constants.MONGO_INPUT_URI, input_uri) \
            .option(constants.MONGO_DATABASE, input_database) \
            .option(constants.MONGO_COLLECTION, input_collection) \
            .load()

        # Write
        writer: DataFrameWriter = input_data.write.mode(output_mode)
        persist_dataframe_to_cloud_storage(writer, args, output_location, output_format, "mongo.gcs.output.")

File: java/src/main/java/com/google/cloud/dataproc/templates/databases/MongoToGCS.java
/*
 * Copyright (C) 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.cloud.dataproc.templates.databases;

import static com.google.cloud.dataproc.templates.util.TemplateConstants.MONGO_GCS_INPUT_COLLECTION;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.MONGO_GCS_INPUT_DATABASE;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.MONGO_GCS_INPUT_URI;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.MONGO_GCS_OUTPUT_FORMAT;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.MONGO_GCS_OUTPUT_LOCATION;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.MONGO_GCS_OUTPUT_MODE;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.SPARK_LOG_LEVEL;

import com.google.cloud.dataproc.templates.BaseTemplate;
import org.apache.commons.lang3.StringUtils;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class MongoToGCS implements BaseTemplate {

  private static final Logger LOGGER = LoggerFactory.getLogger(MongoToGCS.class);

  private final String inputCollection;
  private final String outputFileFormat;
  private final String outputFileLocation;
  private final String outputMode;
  private final String inputUri;
  private final String inputDatabase;
  private final String sparkLogLevel;

  public MongoToGCS() {
    inputCollection = getProperties().getProperty(MONGO_GCS_INPUT_COLLECTION);
    outputFileFormat = getProperties().getProperty(MONGO_GCS_OUTPUT_FORMAT);
    outputFileLocation = getProperties().getProperty(MONGO_GCS_OUTPUT_LOCATION);
    outputMode = getProperties().getProperty(MONGO_GCS_OUTPUT_MODE);
    inputUri = getProperties().getProperty(MONGO_GCS_INPUT_URI);
    inputDatabase = getProperties().getProperty(MONGO_GCS_INPUT_DATABASE);
    sparkLogLevel = getProperties().getProperty(SPARK_LOG_LEVEL);
  }

  @Override
  public void runTemplate() {

    // Initialize Spark Session

    SparkSession spark = null;
    spark = SparkSession.builder().appName("Mongo To GCS").getOrCreate();

    // Set log level
    spark.sparkContext().setLogLevel(sparkLogLevel);

    // Read Data
    Dataset<Row> inputData =
        spark
            .read()
            .format("com.mongodb.spark.sql.DefaultSource")
            .option("spark.mongodb.input.uri", inputUri)
            .option("spark.mongodb.input.database", inputDatabase)
            .option("spark.mongodb.input.collection", inputCollection)
            .load();

    // Write Data
    inputData.write().format(outputFileFormat).mode(outputMode).save(outputFileLocation);
  }

  @Override
  public void validateInput() {
    if (StringUtils.isAllBlank(inputCollection)
        || StringUtils.isAllBlank(outputFileFormat)
        || StringUtils.isAllBlank(outputFileLocation)
        || StringUtils.isAllBlank(inputUri)
        || StringUtils.isAllBlank(inputDatabase)) {
      LOGGER.error(
          "{},{},{},{},{} are required parameter. ",
          MONGO_GCS_INPUT_COLLECTION,
          MONGO_GCS_OUTPUT_FORMAT,
          MONGO_GCS_OUTPUT_LOCATION,
          MONGO_GCS_INPUT_URI,
          MONGO_GCS_INPUT_DATABASE);
      throw new IllegalArgumentException(
          "Required parameters for Mongo to GCS not passed. "
              + "Set mandatory parameter for Mongo to GCS template "
              + "in resources/conf/template.properties file.");
    }
  }
}
